### PR TITLE
Integration test overhaul

### PR DESCRIPTION
**Pull request by Ethan-Arrowood**
## Integration Test Overhaul

This PR introduces a comprehensive overhaul of Harper's integration testing infrastructure, enabling concurrent test execution for faster feedback cycles and improved developer productivity.

### Overview of Changes

**New Integration Test Framework:**
- Created a new integration test framework based on Node.js Test Runner with full support for concurrent test execution
- Implemented loopback address pooling to enable true test isolation - each test file gets its own Harper instance on a unique network address
- Added comprehensive utilities for Harper lifecycle management (`setupHarper`, `teardownHarper`) with automatic resource cleanup
- Introduced new test structure requirements: test files must be **independent**, **hermetic**, and **deterministic**

**Documentation:**
- Added extensive documentation in [`integrationTests/README.md`](integrationTests/README.md) covering test writing guidelines, execution strategies, and best practices
- Created detailed utility documentation in [`integrationTests/utils/README.md`](integrationTests/utils/README.md)
- Updated [`CONTRIBUTING.md`](CONTRIBUTING.md) with new integration test setup requirements

**CI/CD Improvements:**
- Replaced single integration test workflow with new parallelized workflow supporting test sharding across multiple runners
- Build artifacts are now cached and reused across test jobs for faster execution
- Added support for selective Node.js version testing and test sharding (1/4, 2/4, 3/4, 4/4)
- Separated legacy API tests from new integration tests with dedicated workflows

**New Test Infrastructure:**
- Implemented loopback address pool manager for concurrent test execution without port conflicts
- Created reusable Harper lifecycle utilities with configurable installation directories and network settings
- Added helper utilities for tar.gz compression, required for application deployment tests
- Included setup script for configuring loopback addresses on macOS and Linux

**Example Tests:**
- Added deployment tests demonstrating the new testing patterns:
  - [`deploy-from-source.test.mts`](integrationTests/deploy/deploy-from-source.test.mts) - Deploy application from compressed source
  - [`deploy-from-github.test.mts`](integrationTests/deploy/deploy-from-github.test.mts) - Deploy application from GitHub repository

**Additional Changes:**
- Updated ESLint config to allow floating promises in test files for async test functions
- Fixed minor issues in logging and user utilities for better compatibility

### How to Review

1. **Start with the documentation** - Read [`integrationTests/README.md`](integrationTests/README.md) to understand the new testing philosophy, requirements, and execution strategies. This provides critical context for all other changes.

2. **Review the utilities** - Check [`integrationTests/utils/README.md`](integrationTests/utils/README.md) and the utility implementations in [`integrationTests/utils/`](integrationTests/utils/) to understand the building blocks available for writing tests.

3. **Examine the test examples** - Look at the deploy tests ([`deploy-from-source.test.mts`](integrationTests/deploy/deploy-from-source.test.mts) and [`deploy-from-github.test.mts`](integrationTests/deploy/deploy-from-github.test.mts)) as reference implementations showing how to use the new framework and utilities.

4. **Review CI/CD changes** - Check [`.github/workflows/integration-tests.yml`](.github/workflows/integration-tests.yml) to understand how tests are parallelized across runners and how sharding works.

5. **Check supporting changes** - Review smaller changes to source files (installer, logging, etc.) that support better test isolation and execution.

### Testing

The new integration tests require loopback address configuration. Run the setup script if needed:

```bash
bash integrationTests/utils/scripts/setup-loopback.sh
```

Then execute tests:

```
npm run test:integration
```

### CODE REVIEW

File: .github/workflows/integration-tests.yml [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v5
+
+      - name: Setup Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@v6
+        with:
+          node-version: ${{ matrix.node-version }}
+          package-manager-cache: false
```
[the remark this answers is not in the document]
Comment (cb1kenobi) [reply .github/workflows/integration-tests.yml]: Hmm, then add a lockfile?

[the remark this answers is not in the document]
Comment (cb1kenobi) [reply .github/workflows/integration-tests.yml]: Makes sense, thank you!

File: server/threads/threadServer.js
Comment (Ethan-Arrowood) [inline server/threads/threadServer.js]: remove the debug logs from this file.

File: .github/workflows/integration-tests.yml [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+jobs:
+  generate-node-version-matrix:
+    name: Generate Node Version Matrix
+    runs-on: ubuntu-latest
+    outputs:
+      node-versions: ${{ steps.set-node-versions.outputs.node-versions }}
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v5
```
Comment (cb1kenobi) [inline .github/workflows/integration-tests.yml]:
```suggestion
        uses: actions/checkout@v6
```

**Review by kriszyp — COMMENTED** [review]
This looks like it will be awesome, excited to see more tests brought in!

File: utility/install/installer.js [hunk 3]
```diff
@@ -113,8 +116,12 @@ async function install() {
 		throw validationError.message;
 	}
 
-	// Check for an existing install of HarperDB.
-	await checkForExistingInstall();
+	ignoreExisting = promptOverride[hdbTerms.INSTALL_PROMPTS.IGNORE_EXISTING] === 'true';
```
[the remark this answers is not in the document]
Comment (kriszyp) [reply utility/install/installer.js]: I thought using `ROOTPATH` with a different folder than existing installation would never interfere or try to use the existing installation? It seems like that has worked for me, and I think that is why I am little confused about the need for an `ignoreExisting` flag?

[the remark this answers is not in the document]
Comment (Ethan-Arrowood) [reply utility/install/installer.js]: > or try to use the existing installation

My understanding is that this was only the case when some cloneNode code paths were executed. `ignoreExisting` is set to `false` by default here https://github.com/HarperFast/harper/blob/067e550005d6ca9453ee0380135fcc86e99fd134/utility/install/installer.js#L74 and exists in the original code.

And then the `setIgnoreExisting` method (https://github.com/HarperFast/harper/blob/067e550005d6ca9453ee0380135fcc86e99fd134/utility/install/installer.js#L619) was only called in specific `cloneNode.js` code paths (that were all removed as part of the oss split). 

So now, nothing is ever setting `ignoreExisting` to `true`.

**Comment by heskew** [comment]
Generally this is pretty awesome and most of the feedback I've thought of would be non-blocking edge case handling and 'nice to haves' (e.g. loopback cleanup - i.e. leave no trace).

This (test utils especially) might be nice for others to use in their own Harper apps and components/plugins. Have you thought about publishing as a separate package?

File: integrationTests/utils/loopbackAddressPool.mts [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+ * @returns The 0-based pool index for this address
+ * @throws {InvalidLoopbackAddressError} If the address format is invalid or out of range
+ */
+function parseLoopbackAddress(address: string): number {
+	const parts = address.split('.');
+	if (parts.length !== 4 || parts[0] !== '127' || parts[1] !== '0' || parts[2] !== '0') {
+		throw new InvalidLoopbackAddressError(address);
+	}
+
+	const index = parseInt(parts[3], 10) - 1;
```
Comment (cb1kenobi) [inline integrationTests/utils/loopbackAddressPool.mts]: You could use a regex instead of splitting. 

```suggestion
	const m = address.match(/^127\.0\.0\.(\d{1,3})$/);
	if (!m?.[1]) {
		throw new InvalidLoopbackAddressError(address);
	}

	const index = Number.parseInt(m[1], 10) - 1;
```

**Review by cb1kenobi — APPROVED** [review]
This all looks really awesome! Great job!

File: integrationTests/utils/loopbackAddressPool.ts [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+ * @returns The 0-based pool index for this address
+ * @throws {InvalidLoopbackAddressError} If the address format is invalid or out of range
+ */
+function parseLoopbackAddress(address: string): number {
+	const parts = address.split('.');
+	if (parts.length !== 4 || parts[0] !== '127' || parts[1] !== '0' || parts[2] !== '0') {
+		throw new InvalidLoopbackAddressError(address);
+	}
+
+	const index = parseInt(parts[3], 10) - 1;
```
Comment (Ethan-Arrowood) [reply integrationTests/utils/loopbackAddressPool.ts]: Personally, I'm not a huge fan of using regex due to its readability. I understand it can be better performance but since this is just a util script I don't think I'm going to make this change.

Comment (cb1kenobi) [reply integrationTests/utils/loopbackAddressPool.ts]: Heh, I was thinking the regex was so simple that it would be _more_ readable than the split version.

**Review by kriszyp — APPROVED** [review]
Great work, let's start using it!
